### PR TITLE
[FEATURE][needs-docs] import/export section of vector layer style

### PR DESCRIPTION
**Pull request by 3nids**
implements https://github.com/qgis/QGIS-Enhancement-Proposals/issues/125

This moves some things to `QgsMapLayer::read/writeCommonStyle` to avoid code duplication in `read/writeLayerXML` and `import/exportNamedStyle`.
Flags have been added to `QgsMapLayer` to represent style categories of what is export in styles.

### CODE REVIEW

File: src/core/qgsmaplayer.h [hunk 1]
```diff
@@ -736,7 +737,8 @@ class CORE_EXPORT QgsMapLayer : public QObject
      * \param errorMsg this QString will be initialized on error
      * during the execution of writeSymbology
      */
-    virtual void exportNamedStyle( QDomDocument &doc, QString &errorMsg SIP_OUT ) const;
+    virtual void exportNamedStyle( QDomDocument &doc, QString &errorMsg SIP_OUT, QgsReadWriteContext &context,
+                                   QgsMapLayerStyle::StyleCategories categories = QgsMapLayerStyle::All ) const;
```
Comment (m-kuhn) [inline src/core/qgsmaplayer.h]: Technically speaking, this is an API break (if someone subclassed and overwrote).
I guess a mapping could be done but it's definitely not trivial.

Question: is this even actively used? Then at least there would be an example on which to base this.

@nyalldawson @wonder-sk (others?)

**Comment by m-kuhn** [comment]
Looks good to me so far :+1:

File: src/core/qgsmaplayer.h (hunk 1 above)
Comment (nyalldawson) [reply src/core/qgsmaplayer.h]: I don't think overriding layer methods is really intended use of the API. If we could, I'd like them to be marked "final" in the vector/raster classes. But sip doesn't like this, so we're forced with this annoying situation.

Anyway, I'm ok with breaking non-intended use of API.

File: src/app/qgisapp.cpp [hunk 2]
```diff
@@ -9730,7 +9731,8 @@ void QgisApp::duplicateLayers( const QList<QgsMapLayer *> &lyrList )
     // duplicate the layer style
     QString errMsg;
     QDomDocument style;
-    selectedLyr->exportNamedStyle( style, errMsg );
+    QgsReadWriteContext context;
+    selectedLyr->exportNamedStyle( style, errMsg, context );
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp]: Why are these changes needed?

File: src/core/qgsmaplayer.cpp [hunk 3]
```diff
@@ -306,20 +302,6 @@ bool QgsMapLayer::readLayerXml( const QDomElement &layerElement,  QgsReadWriteCo
     }
   }
 
-  // use scale dependent visibility flag
```
Comment (nyalldawson) [inline src/core/qgsmaplayer.cpp]: These changes - while desirable - make me nervous. There's potentially signals getting thrown in the moved code and the change of property restoration order here may have unintentional tricky side effects. I don't think that should block this refactoring... But proceed with extreme caution!

File: src/app/qgisapp.cpp (hunk 2 above)
Comment (3nids) [reply src/app/qgisapp.cpp]: exportNamedStyle was currently using a context created in QgsMapLayer. I think it should be created from the app to be able to report messages back.
Even if not used at the moment, this makes the code as the methods were initially thought.

**Comment by 3nids** [comment]
image

**Comment by saberraz** [comment]
Nice feature!

IMO it will be better to have dedicated buttons on toolbar for copy and pasting style. It is very nested and not very easy to use in this form.

How about having a pint-brush type icon (similar to Word processing and spreadsheets):
https://support.content.office.net/en-us/media/12598786-773c-4e92-be5f-88d766f115ec.jpg
with drop-down options.

**Comment by 3nids** [comment]
@saberraz thanks. There is more details in the QEP. There will be a dialog for advanced usage. I don't think this feature is worth its dedicated icon (I will also implement entries in the layer menu).

File: src/app/qgsapplayertreeviewmenuprovider.cpp [hunk 4]
```diff
@@ -316,11 +316,51 @@ QMenu *QgsAppLayerTreeViewMenuProvider::createContextMenu()
...
+        if ( layer->type() == QgsMapLayer::VectorLayer )
+        {
+          QMenu *copyStyleMenu = menuStyleManager->addMenu( tr( "Copy Style" ) );
+          copyStyleMenu->setToolTipsVisible( true );
+          QList<QgsMapLayer::StyleCategory> categories = qgsEnumMap<QgsMapLayer::StyleCategory>().keys();
+          categories.move( categories.indexOf( QgsMapLayer::AllCategories ), 0 ); // move All categories to top
+          for ( QgsMapLayer::StyleCategory category : categories )
+          {
+            QgsMapLayer::ReadableStyleCategory readableCategory = QgsMapLayer::readableStyleCategory( category );
+            QAction *copyAction = new QAction( readableCategory.icon(), readableCategory.name() );
```
Comment (nyalldawson) [inline src/app/qgsapplayertreeviewmenuprovider.cpp]: this is leaked

File: src/app/qgsapplayertreeviewmenuprovider.cpp [hunk 5]
```diff
@@ -316,11 +316,51 @@ QMenu *QgsAppLayerTreeViewMenuProvider::createContextMenu()
...
+            copyAction->setToolTip( readableCategory.toolTip() );
+            connect( copyAction, &QAction::triggered, this, [ = ]() {app->copyStyle( layer, category );} );
+            copyStyleMenu->addAction( copyAction );
+            if ( category == QgsMapLayer::AllCategories )
+              copyStyleMenu->addSeparator();
+          }
+        }
+        else
+        {
+          menuStyleManager->addAction( tr( "Copy Style" ), app, SLOT( copyStyle() ) );
```
Comment (nyalldawson) [inline src/app/qgsapplayertreeviewmenuprovider.cpp]: would be better to split this call to avoid the old style connection required by the shortcut addAction method

File: src/core/qgsmaplayer.cpp [hunk 6]
```diff
@@ -67,6 +67,51 @@ QString QgsMapLayer::extensionPropertyType( QgsMapLayer::PropertyType type )
...
 }
 
+QgsMapLayer::ReadableStyleCategory QgsMapLayer::readableStyleCategory( QgsMapLayer::StyleCategory category )
+{
+  switch ( category )
+  {
+    case LayerConfiguration:
+      return ReadableStyleCategory( tr( "Layer Configuration" ),
+                                    QgsApplication::getThemeIcon( QStringLiteral( "/propertyicons/system.svg" ) ),
+                                    tr( "Flags, display expression, read-only" ) );
```
Comment (nyalldawson) [inline src/core/qgsmaplayer.cpp]: "Flags" will mean nothing to users. Better to be explicit here.

File: src/core/qgsmaplayer.cpp [hunk 7]
```diff
@@ -67,6 +67,51 @@ QString QgsMapLayer::extensionPropertyType( QgsMapLayer::PropertyType type )
...
+      return ReadableStyleCategory( tr( "Actions" ),
+                                    QgsApplication::getThemeIcon( QStringLiteral( "/propertyicons/action.svg" ) ) );
+    case MapTips :
+      return ReadableStyleCategory( tr( "Map Tips" ),
+                                    QgsApplication::getThemeIcon( QStringLiteral( "/propertyicons/display.svg" ) ) );
+    case Diagrams :
+      return ReadableStyleCategory( tr( "Diagrams" ),
+                                    QgsApplication::getThemeIcon( QStringLiteral( "/propertyicons/diagram.svg" ) ) );
+    case AttributeTable :
+      return ReadableStyleCategory( tr( "Attribute Table" ),
```
Comment (nyalldawson) [inline src/core/qgsmaplayer.cpp]: "Attribute Table Settings"?

File: src/core/qgsmaplayer.cpp [hunk 8]
```diff
@@ -67,6 +67,51 @@ QString QgsMapLayer::extensionPropertyType( QgsMapLayer::PropertyType type )
...
+  switch ( category )
+  {
+    case LayerConfiguration:
+      return ReadableStyleCategory( tr( "Layer Configuration" ),
+                                    QgsApplication::getThemeIcon( QStringLiteral( "/propertyicons/system.svg" ) ),
+                                    tr( "Flags, display expression, read-only" ) );
+    case Symbology :
+      return ReadableStyleCategory( tr( "Symbology" ),
+                                    QgsApplication::getThemeIcon( QStringLiteral( "/propertyicons/symbology.svg" ) ) );
+    case Labels :
```
Comment (nyalldawson) [inline src/core/qgsmaplayer.cpp]: Should 3D be it's own category? (I think yes)

File: src/core/qgsmaplayer.h [hunk 9]
```diff
@@ -133,6 +135,52 @@ class CORE_EXPORT QgsMapLayer : public QObject
...
     Q_FLAG( LayerFlags )
 
+    /**
+     * Categories of style to distinguish appropriate sections for import/export
+     * \since QGIS 3.4
+     */
+    enum StyleCategory
+    {
+      LayerConfiguration = 1 << 0, //!< Flags, display expression, read-only
+      Symbology          = 1 << 1,
```
Comment (nyalldawson) [inline src/core/qgsmaplayer.h]: Can you add dox for all these? I realise some are self explanatory, but it's the precedent

File: src/core/qgsmaplayer.h [hunk 10]
```diff
@@ -133,6 +135,52 @@ class CORE_EXPORT QgsMapLayer : public QObject
...
+    Q_FLAG( StyleCategories )
+
+    /**
+     * Style category with its name and its tooltip translated and readable
+     * \since QGIS 3.4
+     */
+    struct ReadableStyleCategory
+    {
+      public:
+        ReadableStyleCategory( const QString &name, const QString &toolTip = QString() )
```
Comment (nyalldawson) [inline src/core/qgsmaplayer.h]: Constructor dox, getter dox missing

File: src/core/qgsmaplayer.cpp [hunk 11]
```diff
@@ -67,6 +67,51 @@ QString QgsMapLayer::extensionPropertyType( QgsMapLayer::PropertyType type )
...
+                                    QgsApplication::getThemeIcon( QStringLiteral( "/mActionOpenTable.svg" ) ) );
+    case Rendering :
+      return ReadableStyleCategory( tr( "Rendering" ),
+                                    QgsApplication::getThemeIcon( QStringLiteral( "/propertyicons/rendering.svg" ) ),
+                                    tr( "Scale visibility, simplify method, opacity" ) );
+    case CustomProperties :
+      return ReadableStyleCategory( tr( "Custom Properties" ),
+                                    QgsApplication::getThemeIcon( QStringLiteral( "/mActionOptions.svg" ) ) );
+    case AllCategories :
+      return ReadableStyleCategory( tr( "All categories" ) );
```
Comment (nyalldawson) [inline src/core/qgsmaplayer.cpp]: All Categories (capitalization)

File: src/app/qgsapplayertreeviewmenuprovider.cpp (hunk 4 above)
Comment (3nids) [reply src/app/qgsapplayertreeviewmenuprovider.cpp]: fixed

**Comment by 3nids** [comment]
some open questions:
* why is QgsMapLayer::exportNamedStyle virtual? does it make any sense? shall I make it non virtual?
* the are read/writeCustomSymbology signals in vector layer:
  * should I embed the emit signals in a test for a category? 
  * if yes, symbology? new category? custom properties???

File: src/app/qgsvectorlayersavestyledialog.h [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+  public slots:
+    void accept();
+
+  private slots:
+    void showDbHelp();
+    void readUiFileContent( const QString &filePath );
+
+  private:
+    QgsVectorLayer *mLayer;
```
Comment (nyalldawson) [inline src/app/qgsvectorlayersavestyledialog.h]: initialize to nullptr

File: src/app/qgsvectorlayersavestyledialog.cpp [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+  connect( mStyleTypeComboBox, qgis::overload<int>::of( &QComboBox::currentIndexChanged ), this, [ = ]( int idx )
+  {
+    Q_UNUSED( idx );
+    QgsVectorLayerProperties::StyleType type = currentStyleType();
+    mSaveToFileWidget->setVisible( type != QgsVectorLayerProperties::DB );
+    mSaveToDbWidget->setVisible( type == QgsVectorLayerProperties::DB );
+    mStyleCategoriesListWidget->setEnabled( type == QgsVectorLayerProperties::QML );
+    mFileWidget->setFilter( type == QgsVectorLayerProperties::QML ? tr( "QGIS Layer Style File (*.qml)" ) : tr( "SLD File (*.sld)" ) );
+  } );
+  mStyleTypeComboBox->addItem( tr( "as QGIS QML style file" ), QgsVectorLayerProperties::QML );
```
Comment (nyalldawson) [inline src/app/qgsvectorlayersavestyledialog.cpp]: HIG guidelines is that combobox entries are title case: "As QGIS QML Style File"

File: src/app/qgsvectorlayersavestyledialog.cpp [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+  QgsSettings settings;
+
+  switch ( currentStyleType() )
+  {
+    case QgsVectorLayerProperties::QML:
+    case QgsVectorLayerProperties::SLD:
+    {
+      if ( mFileWidget->filePath().isEmpty() )
+      {
+        QMessageBox::information( this, tr( "Save Style in file" ), tr( "Select a file destination." ) );
```
Comment (nyalldawson) [inline src/app/qgsvectorlayersavestyledialog.cpp]: "Save Style in File"

Comment (nyalldawson) [reply src/app/qgsvectorlayersavestyledialog.cpp]: (sorry... just getting in before @DelazJ ;) )

**Review by nyalldawson — COMMENTED** [review]
This is super nice to have! Lovely stuff :tada:

**Comment by nyalldawson** [comment]
Minor thing looking at that screenshot - I think it'd work better if "Categories" was a group box, instead of the centered label on the left

+---Categories ------------------------+
|  [x] Layer Configuration              |

etc

**Comment by nirvn** [comment]
@3nids , great job, this'll be really useful.

**Comment by signedav** [comment]
@3nids Nice :slightly_smiling_face:

**Comment by nyalldawson** [comment]
Merged because @3nids is :sleeping:

File: src/app/qgsvectorlayersavestyledialog.cpp (hunk 14 above)
Comment (DelazJ) [reply src/app/qgsvectorlayersavestyledialog.cpp]: :smiley:

**Comment by DelazJ** [comment]
Looks like the merge commit misses the appropriate tags to generate commits in docs (and changelog?). I can create one in docs but if changelog misses it, someone has to do it (still not sure whether changelog issues are automated or not)

File: src/core/qgsmaplayer.h (hunk 1 above)
Comment (nyalldawson) [reply src/core/qgsmaplayer.h]: If it works, I'm more than happy to use it!
